apps/language_server/lib/language_server/providers/folding_range/heredoc.ex

```diff
@@ -0,0 +1,54 @@
+defmodule ElixirLS.LanguageServer.Providers.FoldingRange.Heredoc do
+  @moduledoc """
+  """
+
+  alias ElixirLS.LanguageServer.Providers.FoldingRange
+
+  @spec provide_ranges([FoldingRange.input()]) :: {:ok, [FoldingRange.t()]}
+  def provide_ranges(%{tokens: tokens}) do
+    ranges =
+      tokens
+      |> group_heredoc_tokens()
+      |> convert_heredoc_groups_to_ranges()
+
+    {:ok, ranges}
+  end
+
+  # The :bin_heredoc token will be either
+  #   - by itself or
+  #   - directly after an :identifier (either :doc or :moduledoc).
+  # :bin_heredoc regions are ended by an :eol token.
+  defp group_heredoc_tokens(tokens) do
+    tokens
+    |> Enum.reduce([], fn
+      {:identifier, _, x} = token, acc when x in [:doc, :moduledoc] ->
+        [[token] | acc]
+
+      {:bin_heredoc, _, _} = token, [[{:identifier, _, _}] = head | tail] ->
+        [[token | head] | tail]
+
+      {:bin_heredoc, _, _} = token, acc ->
+        [[token] | acc]
+
+      {:eol, _, _} = token, [[{:bin_heredoc, _, _} | _] = head | tail] ->
+        [[token | head] | tail]
+
+      _, acc ->
+        acc
+    end)
+  end
+
+  defp convert_heredoc_groups_to_ranges(heredoc_groups) do
+    heredoc_groups
+    |> Enum.map(&FoldingRange.Helpers.first_and_last_of_list/1)
```

Comment:
those maps can be combined to save one iteration

Reply:
Fixed by: 6a921d9